moderntribe/square-one
# Feature/composer performance improvements

This PR greatly improves a raw square one's composer installation, so first some profiling.

I remove all cache folders, vendor folders, the wp/ folder and plugins installed via composer in wp-content/plugins before profiling. Profiling was done via editing start.sh to pass `--profile` to the composer install command. There is some slight deviation from pasting my github key each time by 1 or so seconds.

**NOTE: Need this tested on OSX**

### sq1 raw composer install on the current master branch

`[11.0MiB/198.78s] Memory usage: 10.96MiB (peak: 34.72MiB), time: 198.78s`

### sq1 raw composer install on this branch

`[10.7MiB/16.64s] Memory usage: 10.7MiB (peak: 12.24MiB), time: 16.64s`

So this is **1094.59%** or **11.94x** faster. This should hopefully translate to Travis builds and will certainly improve `composer update` significantly as well.

## Testing Instructions

- Test with a freshly cloned square-one (temporarily rename your `image: tribe-phpfpm:7.2-rev1` to something unique to generate a new build)
- Test composer update
- Test composer install with no composer.lock
- Test changing advanced custom fields
- **Note:** Gravity Forms only supplies its latest version, so it will always download the latest.

Always look at `composer.lock` and make sure the proper URL is still there and not the AWS URL, e.g.

```
"url": "https://www.gravityhelp.com/wp-content/plugins/gravitymanager/api.php?op=get_plugin&slug=gravityforms&key={%WP_PLUGIN_GF_KEY}#v2.4.16.7"
```

Check ACF as well
```
"url": "https://connect.advancedcustomfields.com/index.php?a=download&p=pro&k={%WP_PLUGIN_ACF_KEY}&t=5.8.7"
```

## Findings

The biggest improvements were the following:

- Installing [hirak/prestissimo](https://github.com/hirak/prestissimo) globally via `start.sh` so the first run benefits from its speed increases.
- Installing `zip`, `unzip`, and `php7.2-zip` in the FPM Dockerfile (make sure you rebuild). This significantly decreases installation time and cache writing.
- Setting `"no-api": true,` in VCS repositories. Using `composer install -vvv` I could see it using the GitHub API to fetch every single branch of each repository, with Panels being by far the slowest. This flag will instead just clone the repository into the cache. 
- Removing `source` from `preferred-install` for moderntribe libraries.
- Setting packagist's URL to `https://packagist.org` in the global config (happens in start.sh)

## Bringing these changes into your current project(s)

- Add `dev/docker/composer/` to your .gitignore
- Merge changes from [dev/docker/phpdocker/php-fpm/Dockerfile](https://github.com/moderntribe/square-one/blob/3b50438d113369e7d9e13155a43a96519f4c4fc9/dev/docker/phpdocker/php-fpm/Dockerfile)
- Merge changes from [dev/docker/docker-compose.yml](https://github.com/moderntribe/square-one/blob/3b50438d113369e7d9e13155a43a96519f4c4fc9/dev/docker/docker-compose.yml). Be careful here to not bring in anything you don't need, important lines are: `- ./composer:/application/.composer`, `- COMPOSER_HOME=/application/.composer` and you may have to bump your php-fpm: image: to ensure it rebuilds the image with the zip binaries.
- Merge changes from [dev/docker/start.sh](https://github.com/moderntribe/square-one/blob/3b50438d113369e7d9e13155a43a96519f4c4fc9/dev/docker/start.sh)
- Review  [composer.json](https://github.com/moderntribe/square-one/blob/3b50438d113369e7d9e13155a43a96519f4c4fc9/composer.json) for changes. Importance on ensuring you're preferring `dist` installations and adding `"no-api": true,` to any VCS repository and updating the Gravity Forms URL to use `https`. Ensure you remove `hirak/prestissimo` as this now gets installed globally via the updated `start.sh`
### Cleanup after migrating
- If everything works, due to changes to the Docker volume mapping you can remove the following files/folders on your local: `dev/docker/composer-cache/`, `dev/docker/composer-config.json` 

I should also note the developer for the gravity forms installer accepted my PR.
